src/zpiv/zpivmodule.h

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2019 The PIVX developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+//
+#ifndef PIVX_ZPIVMODULE_H
+#define PIVX_ZPIVMODULE_H
+
+#include "libzerocoin/bignum.h"
+#include "libzerocoin/Denominations.h"
+#include "libzerocoin/CoinSpend.h"
+#include "libzerocoin/Coin.h"
+#include "libzerocoin/SpendType.h"
+#include "primitives/transaction.h"
+#include "script/script.h"
+#include "serialize.h"
+#include "uint256.h"
+#include <streams.h>
+#include <utilstrencodings.h>
+#include "zpiv/zerocoin.h"
+#include "chainparams.h"
+
+class PublicCoinSpend : public libzerocoin::CoinSpend{
+public:
+
+    PublicCoinSpend(libzerocoin::ZerocoinParams* params):pubCoin(params){};
+
+    PublicCoinSpend(libzerocoin::ZerocoinParams* params,
+            CBigNum serial, CBigNum randomness, CPubKey pubkey, std::vector<unsigned char> vchSig):pubCoin(params){
+        this->coinSerialNumber = serial;
+        this->randomness = randomness;
+        this->vchSig = vchSig;
+        this->pubkey = pubkey;
+    };
+
+    template <typename Stream>
+    PublicCoinSpend(
+            libzerocoin::ZerocoinParams* params,
+            Stream& strm):pubCoin(params){
+        strm >> *this;
+    }
+
+    uint8_t getVersion() const { return libzerocoin::PrivateCoin::PUBKEY_VERSION; }
+
+    virtual bool HasValidSerial(libzerocoin::ZerocoinParams* params) const override;
+    virtual bool HasValidSignature() const override ;
+    virtual const uint256 signatureHash() const override { return ptxHash; }
+    libzerocoin::SpendType getSpendType() const { return libzerocoin::SpendType::SPEND; }
+
+    CBigNum randomness;
+    // prev out values
+    uint256 txHash = 0;
+    unsigned int outputIndex = -1;
+    libzerocoin::PublicCoin pubCoin;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(coinSerialNumber);
+        READWRITE(randomness);
+        READWRITE(pubkey);
+        READWRITE(vchSig);
+    }
+};
+
+
+class ZPIVModule {
+
+public:
+    ZPIVModule(){}
```

Comment:
why not using static methods here?
It would not require a concrete instance of this class to be defined as member in CWallet.
I think even making the first two non-member functions in the namespace would do.

For the last one, `validateInput`,that flow imo should be part of a `Verify` method member of `PublicCoinSpend` (overriding the one in `CoinSpend`)